### PR TITLE
Point at match discriminant on type error in match arm pattern

### LINKED ISSUE #57279: Point at match discriminant on type error in match arm pattern
_From https://www.reddit.com/r/rust/comments/abjzi7/expected_usize_found_enum_stdresultresult/_

When encountering [code like the following](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7353fd12938bb6b730a8ee0f4303a77a), where a match arm pattern is not the expected type, point at the match expression to provide more context

```
fn main() {
    let a: usize = 3;
    let b: usize = 3;
    let temp: usize = match a + b {
        Ok(num) => num,
        Err(_) => 0,
    };
}
```

```
error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
4 |    let temp: usize = match a + b {
  |                            ----- this `match` expression is of type `usize`
5 |         Ok(num) => num,
  |         ^^^^^^^ expected usize, found enum `std::result::Result`
  |
  = note: expected type `usize`
             found type `std::result::Result<_, _>`

error[E0308]: mismatched types
 --> src/main.rs:6:9
  |
4 |    let temp: usize = match a + b {
  |                            ----- this `match` expression is of type `usize`
5 |         Ok(num) => num,
6 |         Err(_) => 0,
  |         ^^^^^^ expected usize, found enum `std::result::Result`
  |
  = note: expected type `usize`
             found type `std::result::Result<_, _>`

```

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
4 |    let temp: usize = match a + b {
  |                            ----- this expression has type `usize`
5 |         Ok(num) => num,
  |         ^^^^^^^ expected usize, found enum `std::result::Result`
  |
  = note: expected type `usize`
             found type `std::result::Result<_, _>`
```
Fix #57279.

### CODE REVIEW

File: src/test/ui/mismatched_types/E0409.stderr [hunk 1]
```diff
@@ -9,6 +9,8 @@ LL |         (0, ref y) | (y, 0) => {} //~ ERROR E0409
 error[E0308]: mismatched types
   --> $DIR/E0409.rs:5:23
    |
+LL |     match x {
+   |           - this match expression evaluates to `_`
```
Comment (estebank) [inline src/test/ui/mismatched_types/E0409.stderr]: This isn't an improvement.

File: src/test/ui/match/match-range-fail.stderr [hunk 2]
```diff
@@ -19,6 +19,8 @@ LL |         10 ..= "what" => ()
 error[E0308]: mismatched types
   --> $DIR/match-range-fail.rs:17:9
    |
+LL |     match 5 {
+   |           - this match expression evaluates to `{integer}`
 LL |         'c' ..= 100 => { }
    |         ^^^^^^^^^^^ expected integer, found char
```
Comment (estebank) [inline src/test/ui/match/match-range-fail.stderr]: This error (and https://github.com/rust-lang/rust/pull/57366/files#diff-eb11ffec25036e56175af396f96ddca0) should have a customized diagnostic that points at the range's components pointing out the difference.

File: src/librustc/infer/error_reporting/mod.rs [hunk 3]
```diff
@@ -487,6 +487,9 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
 
     fn note_error_origin(&self, err: &mut DiagnosticBuilder<'tcx>, cause: &ObligationCause<'tcx>) {
         match cause.code {
+            ObligationCauseCode::MatchExpressionArmPattern { span, ty } => {
+                err.span_label(span, format!("this match expression evaluates to `{}`", ty));
```
Comment (varkor) [inline src/librustc/infer/error_reporting/mod.rs]: Perhaps:
``
this match expression has type `{}`
```
is slightly clearer, as "evaluates to" indicates something about the value.

**Comment by michaelwoerister** [comment]
@varkor, would you be up for giving this a full review? Since my review backlog has grown a second page over the break, I'm glad about anything that I can hand off `:)`

**Comment by varkor** [comment]
r? @varkor

I'll take a proper look soon.

**Comment by michaelwoerister** [comment]
Thanks a lot, @varkor!

File: src/librustc_typeck/check/_match.rs [hunk 4]
```diff
@@ -29,12 +29,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         pat: &'gcx hir::Pat,
         mut expected: Ty<'tcx>,
         mut def_bm: ty::BindingMode,
-        is_arg: bool)
```
Comment (varkor) [inline src/librustc_typeck/check/_match.rs]: The comment on `check_pat_walk` is outdated: be helpful to describe `match_discrim_span` for those without context.

**Review by varkor — APPROVED** [review]
Looks good! r=me with the updated comment.

**Comment by pietroalbini** [comment]
@bors retry
AppVeyor... what's wrong with you today?

**Comment by estebank** [comment]
rebased to get bors unstuck

@bors r=varkor

**Pull request merged**